docs/hands-on-learning/advanced/module-2-scanning-recipes.md

```diff
@@ -0,0 +1,83 @@
+* See how to manage shared state across recipe phases using an accumulator.
+
+### Steps
+
+1. With the [`rewrite-recipe-starter`](https://github.com/moderneinc/rewrite-recipe-starter) still open in IntelliJ, open the `AppendToReleaseNotes` recipe.
+   * You can find this recipe in [src/main/java/com/yourorg/AppendToReleaseNotes.java](https://github.com/moderneinc/rewrite-recipe-starter/blob/main/src/main/java/com/yourorg/AppendToReleaseNotes.java).
+2. Review the `AppendToReleaseNotes` recipe and notice the differences from recipes you may have seen before.
+   * Note that it extends `ScanningRecipe` rather than just `Recipe`, and that there is an `Accumulator` member defined to keep track of the data needed by the visitor.
+   * Also, in addition to the `getVisitor()` method, there are three additional overrides: `generate()` and `getScanner()`. All three methods also define an `Accumulator` parameter. There is also an overridden `getInitialValue()` method that returns a new `Accumulator`.
+3. Now open the unit tests for `AppendToReleaseNotes`.
+   * You can find them in [src/test/java/com/yourorg/AppendToReleaseNotesTest.java](https://github.com/moderneinc/rewrite-recipe-starter/blob/main/src/test/java/com/yourorg/AppendToReleaseNotesTest.java).
+4. Review the unit tests to see some of the specifics about testing a scanning recipe.
+   * Notice how `@Test void createNewReleaseNotes() { ... }` uses `org.openrewrite.test.SourceSpecs.text(java.lang.String, java.lang.String, ...)` to provide a before and after text block, with the third parameter using `spec -> spec.path(Path.of("RELEASE.md")` to set the path where the source file either exists or should be created.
+   * The before text block is `null` to indicate that the file does not exist initially. Conversely, you can pass in `null` as the second argument to indicate that the file should be deleted.
+
+## Takeaways
+
+* Scanning recipes enable inspection and generation across multiple files.
+* A custom class can be used as an accumulator to allow state to be shared among the phases of a scanning recipe.
+* Tests for scanning recipes can simulate creation or modification of files using null or empty strings.
+
+
+## Exercise 2b: Write a scanning recipe
+
+In this exercise, you'll write a scanning recipe to find any comments in Java source files that contain `TODO` and track them all together in a separate markdown file.
+
+### Goals for this exercise
+
+* Learn how to use a scanning recipe to collect data across source files.
+* Understand how to use different visitor types for analyzing and generating files.
+* See how scanning recipes support conditional file creation based on project structure.
+
+### Steps
+
+1. Open the unit test [src/test/java/com/yourorg/TrackJavaTodosTest.java](https://github.com/moderneinc/rewrite-recipe-starter/blob/main/src/test/java/com/yourorg/TrackJavaTodosTest.java) in IntelliJ IDEA.
```

Comment:
This link and all of the other links around TrackJavaTodos result in 404s for me.